File: java/src/test/java/test/org/voidland/concurrent/queue/BlownQueueTest.java
package test.org.voidland.concurrent.queue;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.voidland.concurrent.queue.BlownQueue;

import com.google.common.util.concurrent.Uninterruptibles;


public abstract class BlownQueueTest
{
	protected static final int MAX_SIZE = 5;
	
	private BlownQueue<Long> queue;
	
	
    @BeforeEach
    public void initQueue()
        throws Exception
    {
        this.queue = this.createQueue(BlownQueueTest.MAX_SIZE);
    }
    
    protected abstract BlownQueue<Long> createQueue(int maxSize)
    	throws Exception;
	
	
	@Test
	public void test()
		throws Exception
	{
		Random randomGen = new Random(System.currentTimeMillis());
		Set<Long> portions = Collections.synchronizedSet(new HashSet<>());
		Set<Long> retrievedPortions = Collections.synchronizedSet(new HashSet<>());
		
		int producerCount = BlownQueueTest.MAX_SIZE * 10;
		int consumerCount = BlownQueueTest.MAX_SIZE * 10;
		List<Thread> producers = new ArrayList<>(producerCount);
		List<Thread> consumers = new ArrayList<>(consumerCount);
		
		CountDownLatch producersLatch = new CountDownLatch(producerCount + 1);
		for (int i = 0; i < producerCount; i++)
		{
			Thread producer = new Thread(() ->
			{
				Set<Long> myPortions = new HashSet<>();
				
				while (myPortions.size() < BlownQueueTest.MAX_SIZE * 10)
				{
					long portion = randomGen.nextLong();
					if (portions.add(portion))
					{
						myPortions.add(portion);
					}
				}
				
				producersLatch.countDown();
				Uninterruptibles.awaitUninterruptibly(producersLatch);
				for (long portion : myPortions)
				{
					try
					{
						BlownQueueTest.this.queue.addPortion(portion);
					}
					catch (InterruptedException e)
					{
						throw new RuntimeException(e);
					}
				}
			});
			producers.add(producer);
			producer.start();
		}
		
		assertEquals(0, this.queue.getSize());
		producersLatch.countDown();
		while (this.queue.getSize() == 0);

		for (int i = 0; i < consumerCount; i++)
		{
			Thread consumer = new Thread(() ->
			{
				while (true)
				{
					Long portion;
					try
					{
						portion = BlownQueueTest.this.queue.retrievePortion();
					}
					catch (InterruptedException e)
					{
						throw new RuntimeException(e);
					}
					if (portion == null)
					{
						break;
					}
					retrievedPortions.add(portion);
				}
			});
			consumers.add(consumer);
			consumer.start();
		}
		
		for (Thread producer : producers)
		{
			Uninterruptibles.joinUninterruptibly(producer);
		}
		
		this.queue.ensureAllPortionsAreRetrieved();
		assertEquals(0, this.queue.getSize());

		this.queue.stopConsumers(consumerCount);
		for (Thread consumer : consumers)
		{
			Uninterruptibles.joinUninterruptibly(consumer);
		}
		
		assertEquals(portions, retrievedPortions);
	}
	
	@Test
	public void testConsumerBlocking()
		throws Exception
	{
		Thread consumer = new Thread(() ->
		{
			try
			{
				BlownQueueTest.this.queue.retrievePortion();
			}
			catch (InterruptedException e)
			{
				throw new RuntimeException(e);
			}
		});
		consumer.start();
		Uninterruptibles.sleepUninterruptibly(1, TimeUnit.MILLISECONDS);
		
		assertEquals(Thread.State.WAITING, consumer.getState());
		
		this.queue.addPortion(0l);
		consumer.join();
	}
	
	@Test
	public void testProducerBlocking()
		throws Exception
	{
		Thread producer = new Thread(() ->
		{
			try
			{
				for (int i = 0; i <= BlownQueueTest.MAX_SIZE; i++)
				{
					BlownQueueTest.this.queue.addPortion(0l);
				}
			}
			catch (InterruptedException e)
			{
				throw new RuntimeException(e);
			}
		});
		producer.start();
		Uninterruptibles.sleepUninterruptibly(1, TimeUnit.MILLISECONDS);
		
		assertEquals(Thread.State.WAITING, producer.getState());
		
		this.queue.retrievePortion();
		producer.join();
	}

	@Test
	public void testControllerBlocking()
		throws Exception
	{
		this.queue.addPortion(0l);
		
		Thread controller = new Thread(() ->
		{
			try
			{
				BlownQueueTest.this.queue.ensureAllPortionsAreRetrieved();
			}
			catch (InterruptedException e)
			{
				throw new RuntimeException(e);
			}
		});
		controller.start();
		Uninterruptibles.sleepUninterruptibly(1, TimeUnit.MILLISECONDS);
		
		assertEquals(Thread.State.WAITING, controller.getState());
		
		this.queue.retrievePortion();
		controller.join();
	}

	@Test
	public void testMisc()
		throws Exception
	{
		assertTrue(this.queue.getMaxSize() <= BlownQueueTest.MAX_SIZE);
        assertThrows(NullPointerException.class, () ->
        {
            this.queue.addPortion(null);
        });
	}
	
	@Test
	public void testMultipleMutexesDeadlock()
		throws Exception
	{
		this.queue = BlownQueue.createConcurrentBlownQueue(1);
		long count = 1000000;
		
		long consumedPortions[] = new long[1];
		consumedPortions[0] = 0;
		
		Thread consumer = null;
		Thread producer = null;
		try
		{
			consumer = new Thread(() ->
			{
				try
				{
					while (this.queue.retrievePortion() != null)
					{
						consumedPortions[0]++;
					}
				}
				catch (InterruptedException e)
				{
				}
			});
			consumer.setDaemon(true);
			consumer.start();
			
			producer = new Thread(() ->
			{
				try
				{
					for (long i = 0; i < count; i++)
					{
						this.queue.addPortion(i);
					}
				}
				catch (InterruptedException e)
				{
				}
			});
			producer.setDaemon(true);
			producer.start();
			
			producer.join(60000);
			assertFalse(producer.isAlive());
			
			this.queue.stopConsumers(1);
			
			consumer.join(60000);
			assertFalse(consumer.isAlive());
		}
		finally
		{
			if (consumer != null)
			{
				consumer.interrupt();
			}
			if (producer != null)
			{
				producer.interrupt();
			}
		}
		
		assertEquals(count, consumedPortions[0]);
	}
}

File: c++/src/test/BlownQueueTest.cpp
#include "voidland/blown_queue/BlownQueue.hpp"

#include <gtest/gtest.h>
#include <memory>
#include <random>
#include <ctime>
#include <climits>
#include <unordered_set>
#include <mutex>
#include <vector>
#include <thread>
#include <latch>
#include <condition_variable>


using namespace voidland::concurrent::queue;


namespace test::voidland::concurrent::queue
{


template <typename E>
class BlownQueueCreator
{
public:
	virtual ~BlownQueueCreator()
	{
	}

	virtual std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) = 0;

	virtual std::unique_ptr<BlownQueue<E>> createBlownQueueWithMinimalSize()
	{
		return this->createBlownQueue(1);
	}
};


#ifdef HAVE_ATOMIC_QUEUE

template <typename E>
class AtomicBlownQueueCreator :
		public BlownQueueCreator<E>
{
public:
	std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) override
	{
		return createAtomicBlownQueue<E>(maxSize);
	}
};

#endif


#ifdef HAVE_MOODYCAMEL_CONCURRENT_QUEUE

template <typename E>
class ConcurrentBlownQueueCreator :
		public BlownQueueCreator<E>
{
public:
	std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) override
	{
		return createConcurrentBlownQueue<E>(maxSize);
	}
};

#endif


#ifdef HAVE_BOOST_LOCKFREE

template <typename E>
class LockfreeBlownQueueCreator :
		public BlownQueueCreator<E>
{
public:
	std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) override
	{
		return createLockfreeBlownQueue<E>(maxSize);
	}
};

#endif


#ifdef HAVE_XENIUM

template <typename E>
class MichaelScottBlownQueueCreator :
		public BlownQueueCreator<E>
{
public:
	std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) override
	{
		return createMichaelScottBlownQueue<E>(maxSize);
	}
};

template <typename E>
class RamalheteBlownQueueCreator :
		public BlownQueueCreator<E>
{
public:
	std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) override
	{
		return createRamalheteBlownQueue<E>(maxSize);
	}
};

template <typename E>
class VyukovBlownQueueCreator :
		public BlownQueueCreator<E>
{
public:
	std::unique_ptr<BlownQueue<E>> createBlownQueue(std::size_t maxSize) override
	{
		return createVyukovBlownQueue<E>(maxSize);
	}

	std::unique_ptr<BlownQueue<E>> createBlownQueueWithMinimalSize() override
	{
		return this->createBlownQueue(2);
	}
};

#endif



template <typename T>
class BlownQueueTest :
		public testing::Test
{
protected:

	inline static const std::size_t MAX_SIZE = 5;

	std::unique_ptr<BlownQueue<long>> queue;


	void SetUp() override
	{
		this->queue = T().createBlownQueue(BlownQueueTest::MAX_SIZE);
	}
};

using BlownQueueCreators = testing::Types<
#ifdef HAVE_ATOMIC_QUEUE
	AtomicBlownQueueCreator<long>
#endif

#if defined(HAVE_ATOMIC_QUEUE)
	,
#endif

#ifdef HAVE_MOODYCAMEL_CONCURRENT_QUEUE
	ConcurrentBlownQueueCreator<long>
#endif

#if defined(HAVE_ATOMIC_QUEUE) || defined(HAVE_MOODYCAMEL_CONCURRENT_QUEUE)
	,
#endif

#ifdef HAVE_BOOST_LOCKFREE
	LockfreeBlownQueueCreator<long>
#endif

#if defined(HAVE_ATOMIC_QUEUE) || defined(HAVE_MOODYCAMEL_CONCURRENT_QUEUE) || defined(HAVE_BOOST_LOCKFREE)
	,
#endif

#ifdef HAVE_XENIUM
	MichaelScottBlownQueueCreator<long>,
	RamalheteBlownQueueCreator<long>,
	VyukovBlownQueueCreator<long>
#endif
>;

TYPED_TEST_SUITE(BlownQueueTest, BlownQueueCreators);


TYPED_TEST(BlownQueueTest, test)
{
	std::default_random_engine generator(time(0));
	std::uniform_int_distribution<long> distribution(LONG_MIN, LONG_MAX);

	std::unordered_set<long> portions;
	std::mutex portionsMutex;

	std::unordered_set<long> retrievedPortions;
	std::mutex retrievedPortionsMutex;

	std::size_t producerCount = BlownQueueTest<TypeParam>::MAX_SIZE * 10;
	std::size_t consumerCount = BlownQueueTest<TypeParam>::MAX_SIZE * 10;
	std::vector<std::thread> producers;
	std::vector<std::thread> consumers;

	std::latch producersLatch(producerCount + 1);
	for (std::size_t i = 0; i < producerCount; i++)
	{
		producers.emplace_back(std::thread
			{
				[this, &distribution, &generator, &portions, &portionsMutex, &producersLatch]
				{
					std::unordered_set<long> myPortions;
					{
						std::lock_guard<std::mutex> portionsMutexLock(portionsMutex);
						while (myPortions.size() < BlownQueueTest<TypeParam>::MAX_SIZE * 10)
						{
							long portion = distribution(generator);
							if (portions.insert(portion).second)
							{
								myPortions.insert(portion);
							}
						}
					}

					producersLatch.count_down();
					producersLatch.wait();
					for (long portion : myPortions)
					{
						this->queue->addPortion(portion);
					}
				}
			});
	}

	ASSERT_EQ(this->queue->getSize(), 0);
	producersLatch.count_down();
	while (this->queue->getSize() == 0);

	for (std::size_t i = 0; i < consumerCount; i++)
	{
		consumers.emplace_back(std::thread
			{
				[this, &retrievedPortions, &retrievedPortionsMutex]
				{
					while (true)
					{
						std::optional<long> portion = this->queue->retrievePortion();
						if (!portion.has_value())
						{
							break;
						}
						{
							std::lock_guard<std::mutex> retrievedPortionsLock(retrievedPortionsMutex);
							retrievedPortions.insert(*portion);
						}
					}
				}
			});
	}

	for (std::thread& producer : producers)
	{
		producer.join();
	}

	this->queue->ensureAllPortionsAreRetrieved();
	ASSERT_EQ(this->queue->getSize(), 0);

	this->queue->stopConsumers(consumerCount);
	for (std::thread& consumer : consumers)
	{
		consumer.join();
	}

	ASSERT_EQ(portions, retrievedPortions);
}

TYPED_TEST(BlownQueueTest, testConsumerBlocking)
{
	std::atomic<bool> consumerDone(false);

	std::thread consumer
		{
			[this, &consumerDone]
			{
				this->queue->retrievePortion();
				consumerDone.store(true);
			}
		};

	std::this_thread::sleep_for(std::chrono::milliseconds(1));

	ASSERT_EQ(consumerDone.load(), false);

	this->queue->addPortion(0l);
	consumer.join();
}

TYPED_TEST(BlownQueueTest, testProducerBlocking)
{
	std::atomic<bool> producerDone(false);

	std::thread producer
		{
			[this, &producerDone]
			{
				for (int i = 0; i <= this->queue->getMaxSize(); i++)
				{
					this->queue->addPortion(0l);
				}
				producerDone.store(true);
			}
		};

	std::this_thread::sleep_for(std::chrono::milliseconds(1));

	ASSERT_EQ(producerDone.load(), false);

	this->queue->retrievePortion();
	producer.join();
}

TYPED_TEST(BlownQueueTest, testControllerBlocking)
{
	std::atomic<bool> controllerDone(false);

	this->queue->addPortion(0l);
	std::thread controller
		{
			[this, &controllerDone]
			{
				this->queue->ensureAllPortionsAreRetrieved();
				controllerDone.store(true);
			}
		};

	std::this_thread::sleep_for(std::chrono::milliseconds(1));

	ASSERT_EQ(controllerDone.load(), false);

	this->queue->retrievePortion();
	controller.join();
}

TYPED_TEST(BlownQueueTest, testMisc)
{
	ASSERT_TRUE(this->queue->getMaxSize() <= BlownQueueTest<TypeParam>::MAX_SIZE);
}

TYPED_TEST(BlownQueueTest, testMultipleMutexesDeadlock)
{
	std::mutex consumerMutex;
	std::condition_variable consumerCondition;
	std::mutex producerMutex;
	std::condition_variable producerCondition;

	this->queue = TypeParam().createBlownQueueWithMinimalSize();
	long count = 1000000;

	long consumedPortions = 0;

	{
		std::unique_lock consumerLock(consumerMutex);
		std::unique_lock producerLock(producerMutex);

		std::thread consumer
			{
				[this, &consumedPortions, &consumerMutex, &consumerCondition]
				{
					while (this->queue->retrievePortion().has_value())
					{
						consumedPortions++;
					}

					{
						std::unique_lock consumerLock(consumerMutex);
						consumerCondition.notify_one();
					}
				}
			};

		std::thread producer
			{
				[this, count, &producerMutex, &producerCondition]
				{
					for (long i = 0; i < count; i++)
					{
						this->queue->addPortion(i);
					}

					{
						std::unique_lock producerLock(producerMutex);
						producerCondition.notify_one();
					}
				}
			};

		std::cv_status producerStatus = producerCondition.wait_for(producerLock, std::chrono::milliseconds(60000));
		ASSERT_EQ(producerStatus, std::cv_status::no_timeout);
		producer.join();

		this->queue->stopConsumers(1);

		std::cv_status consumerStatus = consumerCondition.wait_for(consumerLock, std::chrono::milliseconds(60000));
		ASSERT_EQ(consumerStatus, std::cv_status::no_timeout);
		consumer.join();
	}

	ASSERT_EQ(consumedPortions, count);
}


}	// namespace test::voidland::concurrent::queue
